Implement connect social accounts functionality

aleksihakli opened the pull request:
Hi @ducheneaut,

I opened an issue #347 a while back and got a few upvotes for implementing social connect functionality in django-rest-auth core.

We've used this connect functionality for quite a while in production and it functions quite nicely. Others have reported it works perfectly as well, so I think this has been validated and proven useful.

If someone wishes to contribute the documentation and / or tests, that would be nice, though the change set is so limited it probably doesn't need much testing. The list and disconnect viewset is probably the most interesting thing that should be tested, and some documentation could be added on the usage of the new components :)

aleksihakli commented:
Hi @maxim-kht, could this be merged into upstream?

maxim-kht commented:
Hi @aleksihakli, thank you for the PR! Could you please provide a little bit more information about social account linking vs. regular social login? We would need some info to add to documentation + for test cases. Is there a relevant piece of documentation in allauth docs? Especially regarding this part https://github.com/Tivix/django-rest-auth/pull/387/files#diff-b8bf644d216766af97c1ed1f0270336dR163

We will review and test those changes. So far I found one issue - this will not work in Python 2: https://github.com/Tivix/django-rest-auth/pull/387/files#diff-b8bf644d216766af97c1ed1f0270336dR162

aleksihakli commented:
Hey @maxim-kht,

the process state is or was not that well documented in allauth, I actually had to do quite a bit of digging of the source code and flows to find and understand the `'connect'` state and where to inject it. It does seem to work though as we are using views like this in production for social account connections for a mobile Facebook SDK in another project.

Sorry about that Python 2 compatibility error! I will amend the PR to fix that.

aleksihakli commented:
The `AuthProcess` state are available here:

https://github.com/pennersr/django-allauth/blob/8e1f0a20646c9af9dc7f7b236e26b351e31a9738/allauth/socialaccount/providers/base.py#L9

We could use `AuthProcess.CONNECT`  symbol for initiating the connect functionality as well, but the `login` and `connect` terms are documented, so I think they won't change and are safe to use as well.

I have fixed the Python 2 compatibility error you found and rebased the  patch set on top of the latest upstream master branch.

aleksihakli commented:
I further refactored the PR and added examples in documentation for implementing Facebook and Twitter connect functionality. The codebase in the current release didn't implement some conditional imports on `allauth.socialaccount` logically, so I had to move some imported code into an `if` block, which might make this PR a bit ugly. Hope that's OK, I think it improves the overall quality of code in this case.

aleksihakli commented on rest_auth/registration/views.py:
```diff
@@ -91,31 +95,96 @@ def post(self, request, *args, **kwargs):
-    example usage for facebook with access_token
-    -------------
-    from allauth.socialaccount.providers.facebook.views import FacebookOAuth2Adapter
+if 'allauth.socialaccount' in settings.INSTALLED_APPS:
```
I moved the `SocialLoginView` definition into the conditional `if` block where the imports reside, I think this cleans up the code slightly.

aleksihakli commented on rest_auth/social_serializers.py:
```diff
@@ -1,75 +1,82 @@
+            return attrs
+
 
-class TwitterLoginSerializer(serializers.Serializer):
```
Moved this into the conditional block as well, no one should be using this without the `allauth.socialaccounts` app installed.

aleksihakli commented on rest_auth/registration/views.py:
```diff
@@ -91,31 +95,96 @@ def post(self, request, *args, **kwargs):
         return Response({'detail': _('ok')}, status=status.HTTP_200_OK)
 
 
-class SocialLoginView(LoginView):
```
I moved the `SocialLoginView` definition into the conditional `if` block where the imports reside, I think this cleans up the code slightly.

maxim-kht commented on rest_auth/registration/serializers.py:
```diff
@@ -186,3 +182,44 @@ def save(self, request):
+                'uid',
+                'last_login',
+                'date_joined',
+                'extra_data',
```
@aleksihakli Should we return extra data here? Depending on provider it might contain some sensitive info that we do not want to expose to frontend my default. For example here's what facebook social account info would display:

```
"extra_data": "{'verified': True, 'locale': 'en_US', 'first_name': 'Maxim', 'id': '{my_facebook_user_id}', 'name': 'Maxim Kukhtenkov', 'link': 'https://www.facebook.com/app_scoped_user_id/{my_facebook_user_id}/', 'updated_time': '2017-08-27T01:17:37+0000', 'email': 'myfacebookemail@example.com', 'last_name': 'Kukhtenkov', 'gender': 'male', 'timezone': -5}"
```

Also notice how it's not parsed as JSON and returned as plain string

aleksihakli replied:
You're right, it's probably better to be a bit stricter. I removed the `extra_data` attribute.

aleksihakli commented:
OK, I'll make have to check for differences in the implementation we are using this functionality in and test the PR through again. I'll report back when the patch set is tested working. Sorry for the hassle.

aleksihakli commented on rest_auth/registration/views.py:
```diff
@@ -5,11 +5,13 @@
 from rest_framework.generics import CreateAPIView
 from rest_framework import status
 
-from allauth.account.adapter import get_adapter
```
I moved this import lower down.

aleksihakli commented on rest_auth/registration/views.py:
```diff
@@ -91,31 +94,98 @@ def post(self, request, *args, **kwargs):
+if 'allauth.socialaccount' in settings.INSTALLED_APPS:
+    from allauth.socialaccount import signals
+    from allauth.socialaccount.models import SocialAccount
+    from allauth.socialaccount.adapters import get_adapter
```
I changed the `get_adapter` to import from the `socialaccount.adapters` path. I suspect this was the error here, but haven't tested E2E yet.

aleksihakli commented:
@maxim-kht I tested the disconnect logic with the new change and it seems to work better with the correct imports. I don't know why the old `get_adapter` import was in the file since it was not used for anything, but with the fixes everything seems OK unless I missed something.

maxim-kht commented:
@aleksihakli Looks like we need both in that file - for login we need DefaultAccountAdapter but for disconnecting social accounts we need DefaultSocialAccountAdapter. I am working on a fix.

maxim-kht commented:
I am going to merge this PR, but will need to refactor few things.

1. We are not using viewsets in the project, and it might not be straightforward to subclass them from third party app, so I refactored it into two separate views - list social accounts and disconnect. I will add more info to documentation

2. Disconnect view currently returns 500 if incorrect social account ID is passed - going to change it to return 404 instead

3. I think we don't really need conditional definition of views and serializers. In order for registration and social auth to work we need allauth installed, and it will be easier to debug, if we define those views in a regular way. The app will raise error if a developer forgets to include allauth, providers, etc. so it will be more explicit what is the cause of issue.

Afterwards I'll add at least some basic unit tests, before releasing new version.

Again, thank you for the PR!

aleksihakli commented:
Thanks for merging! The refactoring sounds good. I thought making those changes as well, but didn't think it was appropriate to suggest it in a simple(ish) PR. To comment quickly:

1. Yeah, that's probably cleaner to just have simple views.
2. That was a bug.
3. Yeah, it's probably a lot cleaner to remove the conditionals.
